File: 408-valid-word-abbreviation/valid-word-abbreviation.java
class Solution {
    public boolean validWordAbbreviation(String word, String abbr) {
        int alen = 0;
        int blen = 0;

        if( word == null || abbr == null){
            return false;
        }

        while(alen < word.length() && blen < abbr.length()){

            if(Character.isDigit(abbr.charAt(blen))){
                if( abbr.charAt(blen) == '0') {
                    return false;
                }

                int total = 0;
                while( blen < abbr.length() && Character.isDigit(abbr.charAt(blen))){
                    total = total* 10 + abbr.charAt(blen) - '0';
                    blen++;
                }

                alen += total;
            }else{
                if ( word.charAt(alen) != abbr.charAt(blen)){
                    return false;
                }
                alen++;
                blen++;
            }
        }

        return blen == abbr.length() && alen == word.length();
    }   
}

File: 408-valid-word-abbreviation/valid-word-abbreviation.py
class Solution:
    def validWordAbbreviation(self, word: str, abbr: str) -> bool:
        count=0
        sub_length = 0
        for ch in abbr:
            if ch.isdigit():
                digit=int(ch)
                if digit==0 and sub_length==0:
                    return False
                else:
                    sub_length=sub_length*10+digit
            else:
                count+=sub_length
                count+=1
                if count>len(word) or ch!=word[count-1]:
                    return False
                sub_length=0
        
        return len(word)==count+sub_length